### PR TITLE
build: Treat all warnings as errors

### PR DESCRIPTION
**Pull request by plexoos**
These changes do not affect how the code is built by default using `cons` but allows one to interrupt compilation when the compiler (GCC 4.8.5) produces a warning.

```
cons EXTRA_CXXFLAGS="-Werror"
```

The source files matching the regex patterns in `mgr/warnoff_dirs.txt` are excluded from the above rule in order to let CI pass with known problems. We hope this will prevent introduction of questionable code in all other directories or new packages.

See issue #392

### CODE REVIEW

File: mgr/ConsDefs.pm [hunk 1]
```diff
@@ -468,7 +468,7 @@
 	$CERNLIB_CPPFLAGS .= " -DCERNLIB_GCC" . $CXX_MAJOR;
         print "CXX_VERSION : $CXX_VERSION MAJOR = $CXX_MAJOR MINOR = $CXX_MINOR\n";
 
-        $CXXFLAGS    = "$XMACHOPT -fPIC -pipe -Wall -Woverloaded-virtual";
+        $CXXFLAGS    = "$XMACHOPT -fPIC -pipe -Wall -Woverloaded-virtual -Werror";
```
Comment (veprbl) [inline mgr/ConsDefs.pm]: Can we *please* limit this to CI. Unconditional `-Werror` can be a huge pain for some of the end users.

Comment (plexoos) [reply mgr/ConsDefs.pm]: yes, we can

Comment (veprbl) [reply mgr/ConsDefs.pm]: So, for implementation, we could define a cons variable like `WARNINGS_AS_ERRORS` and make this line conditional on it. What do you think?

Comment (plexoos) [reply mgr/ConsDefs.pm]: For now, I tried to use the known to cons variable `EXTRA_CXXFLAGS`. So, for pull request builds we can run it as:
```
cons EXTRA_CXXFLAGS="-Werror"
```

**Comment by plexoos** [comment]
Ready for review

**Comment by plexoos** [comment]
Any comments from the code owners?

**Review by genevb — COMMENTED** [review]
To clarify, this allows pre-existing warnings in old codes to not trigger an alert in CI, while new codes will be monitored for such warnings and either fixed or added to this exclusion list?

Is there a clear (documented?) place to make change that would allow someone compiling the whole library to see the warnings for all codes again?

**Comment by plexoos** [comment]
Excellent questions, Gene. The answer to the first question is yes, that is how it is intended to work. With only one comment that the goal should be to shrink the exclusion list rather than adding new codes with questionable constructs. Warnings detect real bugs #431 

For the second one, no special action needs to be taken, the warnings for the excluded paths will still appear in the build log, they just won't fail CI.

**Comment by plexoos** [comment]
If no more comments can we merge?

File: Dockerfile [hunk 2]
```diff
@@ -7,6 +7,8 @@ ARG compiler=gcc485
 
 FROM ghcr.io/star-bnl/star-spack:v0.2.3-${starenv}-${compiler}
 
+ARG compiler
```
Comment (veprbl) [inline Dockerfile:10]: There is an `ARG compiler=gcc485` above. Do we need this line?

**Review by veprbl — APPROVED** [review]

File: Dockerfile (hunk 2 above)
Comment (plexoos) [reply Dockerfile:10]: Yes, it's a feature. That is how you reuse a variable set before the FROM command

Comment (veprbl) [reply Dockerfile:10]: Good to know!

**Comment by plexoos** [comment]
@klendathu2k your last word 🥺

**Review by klendathu2k — APPROVED** [review]
Should be fine, and if/when it breaks the build we will fix.

**Comment by DanielWielanek** [comment]
Actually is a funny story.
For over two weeks, I had a problem with getting only the list of modified files. The script that worked on my local computer didn't work on GitHub servers. Then I found this plugin. As far as I remember, I was not able to connect both plugins correctly (I do not remember why). However, when I removed the plugin for getting the list of modified files, surprisingly, I noticed that the older script is working now. I suppose that by accident, I enabled some flags during tests  :)
So in principle, this plugin actually helped me :)

**Comment by plexoos** [comment]
Cool story, wrong thread 🤣 

Moved to https://github.com/star-bnl/star-sw/pull/436#issuecomment-1344833413

**Pull request merged**